jdaviz/configs/specviz/plugins/unit_conversion/unit_conversion.py

```diff
@@ -159,47 +229,126 @@ def _on_spectral_unit_changed(self, *args):
+                if check_if_unit_is_per_solid_angle(current_y):
+                    self.flux_or_sb.selected = 'Flux'
 
+                untranslatable_units = self.untranslatable_units()
+                # disable translator if flux unit is untranslatable,
+                # still can convert flux units, this just disables flux
+                # to surface brightnes translation for units in list.
+                if flux_or_sb in untranslatable_units:
+                    self.can_translate = False
+                else:
+                    self.can_translate = True
+
+            elif arg['name'] == 'sb_unit_selected':
+                if data_collection_unit == 'Flux':
+                    # when the configuration is Specviz, translation is not currently supported.
+                    # If in Cubeviz, all spectra pass through Spectral Xxtraction plugin and will
+                    # have a scale factor assigned in the metadata, enabling translation.
+                    raise ValueError(
+                        "Unit translation between Flux and Surface Brightness "
+                        "is not supported in Specviz."
+                    )
+                flux_or_sb = self.sb_unit.selected
+                self.can_translate = True
+                # update flux or surface brightness dropdown if necessary
+                if not check_if_unit_is_per_solid_angle(current_y):
+                    self.flux_or_sb.selected = 'Surface Brightness'
+            else:
+                return
+
+        yunit = _valid_glue_display_unit(flux_or_sb, self.spectrum_viewer, 'y')
+
+        # on instantiation, we set determine flux choices and selection
+        # after surface brightness
+        if not self.flux_unit.choices:
+            return
+        if self.spectrum_viewer.state.y_display_unit != yunit:
             self.spectrum_viewer.state.y_display_unit = yunit
             self.hub.broadcast(GlobalDisplayUnitChanged('flux',
-                                                        self.flux_or_sb_unit.selected,
+                                                        flux_or_sb,
                                                         sender=self))
+            self.spectrum_viewer.reset_limits()
 
-    # Ensure first dropdown selection for Flux/Surface Brightness
-    # is in accordance with the data collection item's units.
-    @observe('show_translator')
-    def _set_flux_or_sb(self, *args):
-        if (self.spectrum_viewer and hasattr(self.spectrum_viewer.state, 'y_display_unit')
-           and self.spectrum_viewer.state.y_display_unit is not None):
-            if u.sr in u.Unit(self.spectrum_viewer.state.y_display_unit).bases:
-                self.flux_or_sb_selected = 'Surface Brightness'
-            else:
-                self.flux_or_sb_selected = 'Flux'
-
-    @observe('flux_or_sb_selected')
+    @observe('flux_or_sb_selected', 'flux_unit_selected', 'sb_unit_selected')
     def _translate(self, *args):
```

Comment:
where is this ever called?  What is expected input for `args`?  It seems that they're expected to be dictionaries with specific keys ('name', etc) but without any type-checking or fallbacks.

Reply:
It is only be called when the observe is triggered. You are right about `args`, I will add checks for arg being a dictionary and to make sure `name `exists within it

Reply:
ah, I think we usually call that `msg` then, and it should only be one (no need to accept all positional arguments and loop just to act on the first)

Reply:
That is so much cleaner and nicer